runtime/compiler/optimizer/HCRGuardAnalysis.cpp

```diff
@@ -88,80 +88,99 @@ bool TR_HCRGuardAnalysis::shouldSkipBlock(TR::Block *block)
+         blockId = currentBlock->getNumber();
          if (shouldSkipBlock(currentBlock))
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _exceptionKillSetInfo[blockId]->setAll(numBits);
             treeTop = currentBlock->getExit();
             }
          continue;
          }
 
-      if (treeTop->getNode()->getOpCode().canRaiseException())
-         {
-         exceptingTTSeen = true;
-         }
-
-      osrNode = NULL;
-      if (comp()->isPotentialOSRPoint(treeTop->getNode(), &osrNode) && !checklist.contains(osrNode))
+      if (comp()->isPotentialOSRPoint(ttNode, &osrNode) && !checklist.contains(osrNode))
          {
          checklist.add(osrNode);
          bool supportsOSR = comp()->isPotentialOSRPointWithSupport(treeTop);
          if (supportsOSR)
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _regularGenSetInfo[blockId]->resetAll(numBits);
             }
-         /*else
-            {
-            TR::Node *node = treeTop->getNode();
-            traceMsg(comp(), "mismatch at n%dn %d:%d %s\n", node->getGlobalIndex(), node->getByteCodeInfo().getCallerIndex(), node->getByteCodeInfo().getByteCodeIndex(),  comp()->signature());
-            }*/
-         if (!supportsOSR && treeTop->getNode()->getByteCodeInfo().getCallerIndex() > -1)
+         else
             {
-            _regularGenSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               *(_exceptionGenSetInfo[currentBlock->getNumber()]) |= *(_regularGenSetInfo[currentBlock->getNumber()]);
+            _regularGenSetInfo[blockId]->setAll(numBits);
+            _regularKillSetInfo[blockId]->resetAll(numBits);
             }
          }
-      else if (treeTop->getNode()->isTheVirtualGuardForAGuardedInlinedCall() && comp()->cg()->supportsMergingGuards())
+      else if (ttNode->isTheVirtualGuardForAGuardedInlinedCall() && comp()->cg()->supportsMergingGuards())
          {
-         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(treeTop->getNode());
+         TR_VirtualGuard *guardInfo = comp()->findVirtualGuardInfo(ttNode);
          if (guardInfo->getKind() != TR_HCRGuard)
             {
-            _regularKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
-            if (exceptingTTSeen)
-               _exceptionKillSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+            _regularKillSetInfo[blockId]->setAll(numBits);
+            _regularGenSetInfo[blockId]->resetAll(numBits);
```

Comment:
In order to support killing only on the inlined path, we have to add logic to the analysis and optimizations using the analysis or come up with another data flow to better support our need. Since the taken side usually only has the call with at most some remat trees, optimizing it gives little benefit. Plus there is few optimization that will and can optimize the taken side. As per discussion with @andrewcraik, we decide to avoid the complication and to require optimizations that can generate fear stay away from the taken side.